### PR TITLE
Merging SpeechLLM development branch

### PR DESCRIPTION
**Pull request by pzelasko**
# What does this PR do ?

Support for EMMETT, VoiceTextBlender, enhanced prompt formatter, and various SpeechLLM improvements.

**Collection**: multimodal

# Changelog 
- Support for multimodal SpeechLLM dataloading and training
- Support for TP and FSDP training of SpeechLLM
- Support for EMMETT https://arxiv.org/abs/2409.13523
- Support for VoiceTextBlender https://arxiv.org/abs/2410.17485
- Data types for various text data formats (NMT, chat, multimodal conversation)
- Enhanced prompt formatter with dynamically registered support for data types
- Documentation and tutorials

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/modules/conformer_encoder.py [hunk 1]
```diff
@@ -671,14 +671,15 @@ def forward_internal(
             return audio_signal, length
 
     def update_max_seq_length(self, seq_length: int, device):
+        # TODO: this sync seems unnecessary, remove or disable properly later
```
Comment (pzelasko) [inline nemo/collections/asr/modules/conformer_encoder.py:674]: @zhehuaichen what should we do about this check? It was commented out for practically all of our speechllm experiments and we never had an issue because of that. OTOH when it's not commented out, we risk running into TP deadlocks. Should we just remove this or add an option?

File: nemo/collections/common/data/prompt_fn.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+from typing import Callable, Type
+
+import torch
+
```
Comment (stevehuang52) [inline nemo/collections/common/data/prompt_fn.py:4]: This file seems to have a lot of overlap with `https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/common/prompts/fn.py`?

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 3]
```diff
@@ -467,148 +707,32 @@ def make_structured_with_schema_warnings(config: DictConfig) -> DictConfig:
     return OmegaConf.merge(default, config)
 
 
-@dataclass
```
Comment (pzelasko) [inline nemo/collections/common/data/lhotse/dataloader.py:470]: Moved to `sampling.py`

File: nemo/collections/common/prompts/canary.py [hunk 4]
```diff
@@ -15,13 +15,12 @@
...
 
 import torch
-from lhotse import CutSet, MonoCut
-from lhotse.cut import MixedCut
+from lhotse import MonoCut
+from lhotse.cut import Cut, MixedCut
 from lhotse.utils import ifnone
 
-from nemo.collections.common.prompts.fn import registered_prompt_format_fn
+from nemo.collections.common.data.prompt_fn import registered_prompt_format_fn
```
Comment (stevehuang52) [inline nemo/collections/common/prompts/canary.py:22]: Why move the file outside of the `prompts` folder?

File: nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py [hunk 5]
```diff
@@ -227,8 +227,8 @@ def add_adapter(self, peft_cfgs: Union[PEFTConfig, List[PEFTConfig]]):
             peft_cfgs: One or more PEFTConfig objects that specify the PEFT method configuration
         """
 
-        if self.cfg.optim.name == "distributed_fused_adam":
-            raise ValueError('distributed_fused_adam is not supported for PEFT. Please use fused_adam')
+        # if self.cfg.optim.name == "distributed_fused_adam":
+        #     raise ValueError('distributed_fused_adam is not supported for PEFT. Please use fused_adam')
```
Comment (pzelasko) [inline nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py:231]: @zhehuaichen should we remove this code or restore before merging?

File: nemo/collections/common/data/prompt_fn.py (hunk 2 above)
Comment (pzelasko) [reply nemo/collections/common/data/prompt_fn.py:4]: Good catch, `prompts/fn.py` was supposed to be removed. Updated.

File: nemo/collections/common/prompts/canary.py (hunk 4 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/canary.py:22]: To avoid a circular dependency between prompt formatter APIs and data type APIs. This way we can define default prompt_format_fns right next to data type such as `SourceTargetTextExample` or `NeMoMultimodalConversation`.

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 1 above)
Comment (titu1994) [reply nemo/collections/asr/modules/conformer_encoder.py:674]: This is necessary if your audio samples increase in length than the max length during training -which is all non speech lm training. Keep it. If you want to avoid hitting this, setup a max sequence length for conformer that is larger than the largest it will ever see during training - then this path is skipped entirely by default.

Comment (pzelasko) [reply nemo/collections/asr/modules/conformer_encoder.py:674]: > your audio samples increase in length than the max length during training -which is all non speech lm training.

The samples increase in size during speech lm training as well. That's why I'm saying this check is not necessary at all. I am aware it fixed some issue in the past, but it looks like this solution must have been masking some other issue that is no longer present in the codebase.

> then this path is skipped entirely by default.

The issue is the `if` condition: it performs an NCCL sync which prevents us from using model parallelism.

Comment (pzelasko) [reply nemo/collections/asr/modules/conformer_encoder.py:674]: Talked with @titu1994 : we'll add a flag here

File: nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py (hunk 5 above)
Comment (zhehuaichen) [reply nemo/collections/nlp/parts/mixins/nlp_adapter_mixins.py:231]: restore

File: nemo/collections/asr/modules/conformer_encoder.py (hunk 1 above)
Comment (zhehuaichen) [reply nemo/collections/asr/modules/conformer_encoder.py:674]: thanks

**Review by zhehuaichen — COMMENTED** [review]
lgtm on multimodal/speech_llm

File: nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py [hunk 6]
```diff
@@ -79,102 +81,108 @@ def __init__(
...
-        get_max_len(fields["input_ids"]), get_max_len(fields["context_ids"]), get_max_len(fields["answer_ids"])
-    )
-    # increase max length to nearest multiple of 4 or 8
+    input_id_maxlen = get_max_len(fields["input_ids"])
+    context_id_maxlen = tokens_to_generate + get_max_len(fields["context_ids"])
+    answer_id_maxlen = get_max_len(fields["answer_ids"])
     if pad_to_max_length:
-        max_length = max_seq_length
-    else:
-        max_length = min(max_seq_length, ceil_to_nearest(max_length, 8))
```
Comment (stevehuang52) [inline nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:159]: the `ceil_to_nearest()` is needed of sequence parallel, although we currently don't support sequence parallel in speechlm, we might need to add this back in the future

Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/data/lhotse_dataset.py:159]: Let's restore it when adding sequence parallel support

File: nemo/collections/multimodal/speech_llm/data/build_dataset.py [hunk 7]
```diff
@@ -147,16 +133,25 @@ def build_speechllm_dataloader(dataset, data_cfg, consumed_samples=0, is_predict
...
                     data_cfg.input_cfg = input_cfg
                     # for getting names
-                    manifest_filepath = [ic.manifest_filepath for ic in input_cfg[0].input_cfg]
+                    manifest_filepath = []
+                    for ic in input_cfg[0].input_cfg:
+                        if hasattr(ic, "manifest_filepath"):
+                            manifest_filepath.append(ic.manifest_filepath)
+                        else:
+                            assert ic.type == "txt_pair"
+                            manifest_filepath.append(ic.target_paths)
```
Comment (stevehuang52) [inline nemo/collections/multimodal/speech_llm/data/build_dataset.py:142]: Is there a reason why we have to use different attribute for `txt_pair`? It looks better if it has the same `manifest_filepath` input as other types

File: nemo/collections/multimodal/speech_llm/modules/perception_modules.py [hunk 8]
```diff
@@ -487,22 +487,23 @@ def forward(
         return_mems=False,
     ):
         assert input_embeds.shape[-1] == encoder_states.shape[-1]
```
Comment (stevehuang52) [inline nemo/collections/multimodal/speech_llm/modules/perception_modules.py:489]: Can we add more debug info here, like printing out the two shapes?

File: nemo/collections/multimodal/speech_llm/modules/perception_modules.py [hunk 9]
```diff
@@ -487,22 +487,23 @@ def forward(
...
+                return_mems=return_mems,
+                return_mems_as_list=False,
+            )
+            if return_mems:
+                extra_outpus = {'decoder_mems_list': y}
+                y = y[-1][:, -input_embeds.shape[1] :]
+            else:
+                extra_outpus = {}
+            y = self.output_proj(y) + input_embeds
+            assert y.shape == input_embeds.shape
```
Comment (stevehuang52) [inline nemo/collections/multimodal/speech_llm/modules/perception_modules.py:508]: Can we add more debug info here, like printing out the two shapes?

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py [hunk 10]
```diff
@@ -892,7 +892,14 @@ def configure_optimizers(self):
                     ]
                 for bucket in buckets:
                     self._optimizer.init_params_bucket(bucket)
-                self._optimizer.init_params_bucket(self.parameters())
+                try:
+                    params = self.parameters(requires_grad_only=True)
+                except TypeError as e:
```
Comment (stevehuang52) [inline nemo/collections/nlp/models/language_modeling/megatron_base_model.py:897]: Could you please put more explanations on this change? What does it do and why is it necessary? I suppose it's related to mixed modality training, but we need more explanations to help other people understand the code

**Review by stevehuang52 — COMMENTED** [review]
Awesome work~! LGTM on my side, but need @zhehuaichen to double check the details since I'm not totally familiar with the recent changes in this PR.

File: nemo/collections/common/prompts/formatter.py [hunk 11]
```diff
@@ -255,6 +262,11 @@ def encode_dialog(self, turns: list[dict]) -> dict[str, torch.Tensor]:
         turn_token_counts = []
         turn_mask_values = []
 
+        if self.INSERT_BOS:
```
Comment (titu1994) [inline nemo/collections/common/prompts/formatter.py:265]: What if there are custom BOS and EOS tokens that are not part of tokenizer.bos/eos? I think some models have multiple tokens to halt / their EOS os more like EOT (llama). Manually put it inside the prompt ?

File: examples/multimodal/speech_llm/conf/modular_audio_gpt_config_eval.yaml [hunk 12]
```diff
@@ -104,7 +104,7 @@ model:
       prompt_template: ${data.train_ds.prompt_template} # don't change, let hydra resolve from saved config
       tokens_to_generate: 512
       log_every_n_steps: 1
-      sample_rate: ${data.train_ds.sample_rate} # don't change, let hydra resolve from saved config
+      sample_rate: 16000 # don't change, let hydra resolve from saved config
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply examples/multimodal/speech_llm/conf/modular_audio_gpt_config_eval.yaml:107]: I think I had trouble with auto-resolution in some context where I read this config. I can revert it.

File: nemo/collections/common/data/lhotse/cutset.py [hunk 13]
```diff
@@ -30,38 +30,93 @@
...
+    """
+    Return the parsing function for a given data type name.
+    Parsing function reads a dataloading config and returns a tuple
+    of lhotse ``CutSet`` and boolean indicating whether we should use
+    iterable dataset (True) or map dataset (False) mechanism ("is tarred").
+    """
+    return KNOWN_DATA_CONFIG_TYPES[data_type_name]
+
+
+def data_type_parser(name: str | list[str]):
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/cutset.py:93]: "Data type" because it is used to read serialized data in some format and parse it into "data types" introduced in this PR. Registry - yes, cutset - not sure... data type reader maybe? to me both sound equally good / bad.

File: nemo/collections/common/data/lhotse/dataloader.py [hunk 14]
```diff
@@ -198,20 +229,220 @@ def get_lhotse_dataloader_from_config(
     The ``tokenizer`` is used both for audio and text datasets for on-the-fly tokenization.
     This allows us to stratify the bucketing by the count of input/output tokens (depending on modality).
     If "prompt_format" is additionally provided in the config, we will also apply a prompt formatter.
+    Note that ``tokenizer`` can be any tokenizer type (e.g. both SentencePiece and Aggregate tokenizers work).
+    """
+    if not isinstance(config, DictConfig):
+        config = OmegaConf.create(config)
+
+    # Providing default value because we haven't filled the config defaults yet.
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/dataloader.py:237]: Yes, we checked that earlier when this was first added (~half year ago). It's not a new feature in this PR, just preserving it for multi-sampler setups.

File: nemo/collections/common/data/lhotse/text_adapters.py [hunk 15]
```diff
@@ -81,17 +191,404 @@ def __post_init__(self):
...
+    return prompt.encode_dialog(
+        [
+            {"role": "user" if turn["from"] == "User" else prompt.OUTPUT_ROLE, "slots": {"message": turn["value"]}}
+            for turn in example.data["conversations"]
+        ]
+    )
+
+
+@dataclass
+class NeMoSFTJsonlAdapter:
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/common/data/lhotse/text_adapters.py:261]: Based on the new section in the documentation that describes each format; we'll also need notebooks with end to end examples for speech-llm models as we discussed offline.

File: nemo/collections/common/prompts/formatter.py (hunk 11 above)
Comment (pzelasko) [reply nemo/collections/common/prompts/formatter.py:265]: Yes, every custom/special token is normally expressed as a regular string in the prompt template. INSERT_BOS/EOS is specifically for LLMs that use sentencepiece tokenizers which have standard BOS/EOS that don't normally exist in the tokenizer's vocab (i.e. no string input generates them and you must insert them programmatically), see: https://github.com/google/sentencepiece/issues/102#issuecomment-397150427

Comment (titu1994) [reply nemo/collections/common/prompts/formatter.py:265]: Ok this needs to be clearly stated in the code itself (maybe I missed it)

File: nemo/collections/multimodal/speech_llm/data/build_dataset.py (hunk 7 above)
Comment (stevehuang52) [reply nemo/collections/multimodal/speech_llm/data/build_dataset.py:142]: @pzelasko could you please clarify this?

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py (hunk 10 above)
Comment (stevehuang52) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py:897]: @pzelasko could you please address this?

File: nemo/collections/multimodal/speech_llm/data/build_dataset.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/multimodal/speech_llm/data/build_dataset.py:142]: yes, followed up offline - TL;DR this is to support an NMT format where source and target translations are stored in separate plain text files

**Review by zhehuaichen — APPROVED** [review]
lgtm on speech_llm collection
Thank you so much!

File: nemo/collections/multimodal/speech_llm/models/modular_models.py
Comment (biscayan) [inline nemo/collections/multimodal/speech_llm/models/modular_models.py:1]: @pzelasko 
Hi, I tested the speechllm code, and I faced an error on this file.
In line 661, output.contiguous() made an error "tuple object has no attribute contiguous"
Output has a tuple format. Can the tuple have contiguous() method?
